server/src/main/scala/org/http4s/server/websocket/WebSocketBuilder.scala

```diff
@@ -52,22 +88,25 @@ object WebSocketBuilder {
           .withEntity("WebSocket handshake failed.")
           .pure[F],
         onClose: F[Unit] = Applicative[F].unit,
-        filterPingPongs: Boolean = true): F[Response[F]] = {
+        filterPingPongs: Boolean = true)(implicit FConcurrent: Concurrent[F]): F[Response[F]] = {
 
       val finalReceive: Pipe[F, WebSocketFrame, Unit] = if (filterPingPongs) {
         _.filterNot(isPingPong).through(receive)
       } else receive
 
-      WebSocketBuilder(send, finalReceive, headers, onNonWebSocketRequest, onHandshakeFailure).onNonWebSocketRequest
+      onNonWebSocketRequest
         .map(
           _.withAttribute(
             websocketKey[F],
-            WebSocketContext(WebSocket(send, finalReceive, onClose), headers, onHandshakeFailure))
+            WebSocketContext(
+              WebSocketSeparatePipe(send, finalReceive, onClose),
+              headers,
+              onHandshakeFailure))
         )
     }
   }
 
-  def apply[F[_]: Applicative]: Builder[F] = new Builder[F]
+  def apply[F[_]: Concurrent]: Builder[F] = new Builder[F]
```

Comment:
This is probably the most controversial change.  It's why this will be a binary-breaking change, even though `F` will almost always be `Concurrent` in practice.

I think we should either upgrade the constraint on the constructor to `Concurrent`, or move the `Applicative` constraint to `buildWithSinglePipe`, the way it is on `build`.  Our builders do more of the former, though I've never convinced myself one style is superior to the other.

Reply:
I eliminated dependency on `Concurrent` by moving `concurrently` construction back to `Http4sWSStage`